Add auto create action

martijnvg opened the pull request:
Currently the `TransportBulkAction` detects whether an index is missing and
then decides whether it should be auto created. The coordination of the
index creation also happens in the `TransportBulkAction` on the coordinating node.

This change adds a new transport action that the `TransportBulkAction` delegates to
if missing indices need to be created. The reasons for this change:
* Auto creation of data streams can't occur on the coordinating node.
  Based on the index template (v2) either a regular index or a data stream should be created.
  However if the coordinating node is slow in processing cluster state updates then it may be
  unaware of the existence of certain index templates, which then can load to the
  `TransportBulkAction` creating an index instead of a data stream. Therefor the coordination of
  creating an index or data stream should occur on the master node. See #55377
* From a security perspective it is useful to know whether index creation originates from the
  create index api or from auto creating a new index via the bulk or index api. For example
  a user would be allowed to auto create an index, but not to use the create index api. The
  auto create action will allow security to distinguish these two different patterns of
  index creation.

This change adds the following new transport actions:
* `AutoCreateAction`, the `TransportBulkAction` redirects to this action and this action will actually create the index (instead of the `TransportCreateIndexAction`). Later via #55377, can improve the `AutoCreateAction` to also determine whether an index or data stream should be created. 

The `create_index` index privilege is also modified, so that if this permission is granted then a user is also allowed to auto create indices. This change does not yet add an `auto_create` index privilege. A future change can introduce this new index privilege or modify an existing index / write index privilege.

martijnvg committed:
Add auto create action

Currently the TransportBulkAction detects whether an index is missing and
then decides whether it should be auto created. The coordination of the
index creation also happens in the TransportBulkAction on the coordinating node.

This change adds a new transport action that the TransportBulkAction delegates to
if missing indices need to be created. The reasons for this change:
* Auto creation of data streams can't occur on the coordinating node.
  Based on the index template (v2) either a regular index or a data stream should be created.
  However if the coordinating node is unaware of certain index templates then the TransportBulkAction
  could create an index instead of a data stream. Therefor the decision of whether an index or
  data stream should be created should happen on the master node. See #55377
* From a security perspective it is useful to know whether index creation originates from the
  create index api or from auto creating a new index via the bulk or index api. For example
  a user would be allowed to auto create an index, but not to use the create index api. The
  auto create action will allow security to distinguish these two different patterns of
  index creation.

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Indices APIs)

martijnvg committed:
take into account date math in index names.

martijnvg committed:
Change request class to implement `IndicesRequest` interface, similar to create index request.

martijnvg committed:
Split index creation part of the AutoCreateAction class to AutoCreateIndexAction.

This is needed, because AutoCreateAction will make the decision whether to auto create an data stream
or index, and AutoCreateIndexAction really auto creates the index. Future change will add
AutoCreateDataStream action. If this logic is in a single action then we can't distingues
whether a data stream or index is being auto created, which is important because a user may be
able to auto create a data stream, but not an index or visa versa.

martijnvg committed:
fixed checkstyle violations

martijnvg committed:
Merge remote-tracking branch 'es/master' into auto_create_action!

martijnvg committed:
iter

martijnvg committed:
Merge remote-tracking branch 'es/master' into auto_create_action!

martijnvg commented:
@elasticmachine run elasticsearch-ci/1

martijnvg committed:
fixed test

martijnvg commented:
@elasticmachine run elasticsearch-ci/docs

martijnvg committed:
Merge remote-tracking branch 'es/master' into auto_create_action!

henningandersen commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -235,36 +238,91 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
+                        Exception suppressed = null;
+                        for (Map.Entry<String, Exception> entry : response.getFailureByNames().entrySet()) {
+                            Exception e = entry.getValue();
+                            if (e == null || ExceptionsHelper.unwrapCause(e) instanceof ResourceAlreadyExistsException) {
```
I think it would be nicer to handle the `ResourceAlreadyExistsException` in the auto-create action.

henningandersen commented on server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java:
```diff
@@ -0,0 +1,244 @@
+                createIndexResponse -> {
+                    // Maybe a bit overkill to ensure visibility of results map across threads...
+                    synchronized (results) {
+                        results.put(name, null);
```
The map is in the response named "failureByNames". I think we should either not add successful indices into it, rename the map or separate the two parts in the response.

henningandersen commented on server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java:
```diff
@@ -0,0 +1,244 @@
+            createIndexRequest.cause(request.getCause());
+            createIndexRequest.masterNodeTimeout(request.masterNodeTimeout());
+            createIndexRequest.preferV2Templates(request.getPreferV2Templates());
+            client.execute(AutoCreateIndexAction.INSTANCE, createIndexRequest, ActionListener.wrap(
```
By delegating to a separate transport action, we risk that the retry on no longer master happens inside that transport action, making any decision we make on cluster state here invalid (or at least potentially stale). I would much prefer to do the cluster state update here.

That way we could also create all the indices and streams in one go, reducing the number of cluster states published.

I think auto-create does not have to be specific to data stream or index. It goes together with the document level privileges which are also agnostic of whether the data ends up in a data stream or an index.

martijnvg replied:
> I think auto-create does not have to be specific to data stream or index. It goes together with the document level privileges which are also agnostic of whether the data ends up in a data stream or an index.

Yes and this is another reason to keep all the auto creation logic in a single action.

henningandersen reviewed (changes requested):
Thanks @martijnvg , I did an initial review and added a few comments, primarily the one we already discussed on collapsing the two transport actions into one.

martijnvg committed:
Merge remote-tracking branch 'es/master' into auto_create_action!

martijnvg committed:
Simplify auto create index logic:
* One action for auto creating indices and in the future also auto create data streams
* The TransportBulkAction just executes the auto create action like the create index action.
* Avoid the complexity of auto creating indices in a single request / cluster state update.
  This adds quite some complexity while the benefits are likely only noticeable in edge cases
  (if many indices get auto created)) Also on the security side, authorization of the auto
  create indices would become much more complex compared to authorization of auto creating
  an index one at a time.

martijnvg committed:
Merge remote-tracking branch 'es/master' into auto_create_action!

martijnvg committed:
removed unused imports and class

martijnvg commented:
I've changed this PR, so that there is now a single auto create action that auto creates an index. This can then later be used to also auto create data streams. The auto create action also now handle only a single index at a time instead of multiple indices. The additional complexity isn't worth the benefit from auto creating multiple indices in a single remote call / cluster state update. Also this would require additional logic in security to deny / allow specific indices from being auto created.

martijnvg committed:
Added integration test for auto index creation and the auto_create index privilege

henningandersen reviewed (approved):
LGTM.

martijnvg commented:
I've added the breaking label, because the privilege `indices:admin/create` can no longer be used  to allow auto creation of indices and `create_index` should be used instead. 

The `indices:admin/create` is a privilege based on an internal transport action and we don't provide backwards compatibility guarantees on transport action names being used as privilege and instead the [named privileges](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-privileges.html#privileges-list-indices) should be used. 

However there is usage of `indices:admin/create` privilege and therefor I'm marking this pr as breaking.
